### ISSUE TITLE
Role proposal: "no-arguments-in-arrow-function"

### ISSUE THREAD

**Issue by gajus** [eslint/eslint#6204]
> The arguments object is an Array-like object corresponding to the arguments passed to a function.

https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/arguments

> Arrow functions do not expose an arguments object to their code: arguments.length, arguments[0], arguments[1], and so forth do not refer to the arguments provided to the arrow function when called.  Instead, arguments is simply a reference to the name in the enclosing scope.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Lexical_arguments

This rule prevents accidental reference to a parent function's `arguments` object when the developer intention is to get parameters of the current function. This rule prevents accidental reference to an undefined `arguments` variable when the developer intention is to get parameters of the current function.

Examples of **incorrect** code for this rule:

```js
() => {
    return arguments;
};
```

```js
function foo () {
    return () => {
        return arguments;
    };
};
```

Examples of **correct** code for this rule:

```js
(...args) => {
    return args;
};
```
```js
function foo () {
    const fooArguments = arguments;

    return () => {
        return fooArguments;
    };
};
```

**Comment by mysticatea** [comment]
```js
() => {
    return arguments;
};
```

This is warned by `no-undef`.

```js
function foo() {
    return () => {
        return arguments;
    };
}
```

This is waned by `prefer-rest-params`.

I'm not sure this is proper for core because there are overlaps.

**Comment by nzakas** [comment]
Yeah, I'm skeptical as well. Arrow functions use an enclosing junction's `arguments`, `this`, and `super`, and that's by design. I think if we did anything, it would have to be for all three. 

I could see where some might like such a rule, just not sure it's right for the core.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
